### PR TITLE
Fix nil pointer panic whens stopping exporters

### PR DESCRIPTION
**Pull request by huikang**
- Exporters not used in any pipeline cause panic during shutting
  down.

Signed-off-by: Hui Kang <kangh@us.ibm.com>

### CODE REVIEW

**Comment by huikang** [comment]
Fixed. Thanks.

File: service/builder/exporters_builder.go [hunk 1]
```diff
@@ -172,7 +174,7 @@ func (eb *ExportersBuilder) buildExporter(
 		// if there are no pipelines associated with the exporter.
 		eb.logger.Warn("Exporter " + config.Name() +
 			" is not associated with any pipeline and will not export data.")
-		return exporter, nil
+		return nil, nil
```
Comment (songy23) [inline service/builder/exporters_builder.go]: Not sure this is what we want in this case. I think even if an exporter is not used in any pipelines, we still want to build and return it as long as it's specified. WDYT? @pjanotti @tigrannajaryan

Comment (pjanotti) [reply service/builder/exporters_builder.go]: It seems strange to return nil and nil error. I think that the natural behavior here is to return error instead of log a warning. What is the current behavior if the exporter is explicitly disabled on the config? If `disabled` is not set to true and the exporter is not connected to any pipeline it seems reasonable to error in that case, if disabled is set to `true`, independent of the internal implementation, it seems reasonable to allow the collector to go ahead with perhaps logging a warning.

Comment (huikang) [reply service/builder/exporters_builder.go]: @pjanotti , the current behavior is that if an exporter is enabled (which is the default) and not associated with any pipeline, exiting the agent will causes a nil pointer panic.

- I added a condition to check if the exporter is enabled in the updated PR.

File: service/builder/exporters_builder_test.go [hunk 2]
```diff
@@ -37,8 +37,9 @@ func TestExportersBuilder_Build(t *testing.T) {
 		Exporters: map[string]configmodels.Exporter{
 			"opencensus": &opencensusexporter.Config{
 				ExporterSettings: configmodels.ExporterSettings{
-					NameVal: "opencensus",
-					TypeVal: "opencensus",
+					NameVal:  "opencensus",
+					TypeVal:  "opencensus",
+					Disabled: false,
```
Comment (tigrannajaryan) [inline service/builder/exporters_builder_test.go]: Why is `Disabled: false` needed? false is zero value for bool.

File: service/builder/exporters_builder.go [hunk 3]
```diff
@@ -168,8 +168,11 @@ func (eb *ExportersBuilder) buildExporter(
 
 	inputDataTypes := exportersInputDataTypes[config]
 	if inputDataTypes == nil {
-		// No data types where requested for this exporter. This can only happen
-		// if there are no pipelines associated with the exporter.
+		if config.IsEnabled() {
+			return nil, fmt.Errorf("Exporter " + config.Name() +
```
Comment (tigrannajaryan) [inline service/builder/exporters_builder.go]: Unclear why this should be an error. It seems that giving a warning (below) should be sufficient.

File: service/builder/exporters_builder.go (hunk 1 above)
Comment (tigrannajaryan) [reply service/builder/exporters_builder.go]: I agree with @songy23 and @pjanotti 
We already give warning on exporter that is not used in a pipeline. I don't see a need to make this an error.
If this causes a panic later we need to fix that panic, not make it an erroneous config.

Comment (huikang) [reply service/builder/exporters_builder.go]: Hi, @tigrannajaryan , thanks for your comment. The panic is caused by having a non-nil exporter with nil `stopFunc` [1], if the exporter is not used by any pipeline.

If I understand @pjanotti 's suggestion correctly, `` If disabled is not set to true and the exporter is not connected to any pipeline it seems reasonable to error in that case``. This is exactly what the PR does.

[1] https://github.com/open-telemetry/opentelemetry-service/blob/f20385bac8e7a3db68a1f8cfeaa3a3edcde9f419/service/builder/exporters_builder.go#L178-L206

File: service/builder/exporters_builder.go (hunk 3 above)
Comment (huikang) [reply service/builder/exporters_builder.go]: As I mentioned earlier, the created eb will have nil stopFunc, which causes panic during process exiting.

**Comment by ccaraman** [comment]
Is it possible to include the call stack in the description? Is this problem with processors not associated with pipelines?

**Comment by huikang** [comment]
Hi, @ccaraman , please see #244. Thanks.

File: service/builder/exporters_builder.go (hunk 3 above)
Comment (tigrannajaryan) [reply service/builder/exporters_builder.go]: @huikang:

I do not see why the panic during shutdown is conflated with the proposed behavior change during configuration loading at startup.

There is a piece of code which panics when it should not. This needs to be fixed without changing unrelated functionality: logging of warning on certain configuration.

If you think that it is better to give an error in this situation instead of giving a warning then that is  best to be discussed separately in a separate issue. For now I do not see good reasons for changing a warning to an error. If you can bring such reasons I will be happy to consider.

Otherwise, I think it is best to fix the panic bug during shutdown without changing the intended behavior of the program during configuration reading, i.e. keep the warning.

In this case the root cause of the panic is that `buildExporter()` function violates the contract of `builtExporter` type. It is expected that every valid `builtExporter` instance must have non-nil `stop` member.

The correct solution is to fix `buildExporter()` function to honour the type contract and supply non-nil `stop` member. This can be done by adding one line at line 175 before return statement:
```go
exporter.stop = func() error { return nil }
```
This should fix the panic bug without changing intended application behavior during startup.

Comment (huikang) [reply service/builder/exporters_builder.go]: @tigrannajaryan , yours is better. My concern was about the behavior of the stop function. Now it seems a simple `func() error { return nil }` should be good enough.

I will update the PR.

Comment (tigrannajaryan) [reply service/builder/exporters_builder.go]: @huikang if you can also add a test that verifies this edge case it would be great.

Comment (tigrannajaryan) [reply service/builder/exporters_builder.go]: Ah, I see, you already did. Thanks!
